### ISSUE TITLE
Allow collection model subclasses to be defined without using protocols

### ISSUE THREAD

**Issue by PerrchicK** [jsonmodel/jsonmodel#398]
An error is thrown after trying to deserialize a typed array in Swift. The type of the array also inherits from JSONModel but the deserialization does not dive into the dictionaries inside the array. So instead of giving an array with objects of JSONModelSubClass it gives an array of NSDictionary objects.

The error:
__"fatal error: NSArray element failed to match the Swift Array Element type"__

So, I fixed that by giving an option to supply a dictionary with array names and their types.
Like JSONKeyMapper is initialized with custom mapping dictionary, I gave the option to implement a method for declaring arrays and types mapping.

I'll upload my fix to this issue's branch and you guys can review it.

Thanks!

**Comment by rishabhtayal** [comment]
+1

**Comment by hanneskl** [comment]
+1

**Comment by renatomedina** [comment]
+1

**Comment by billinghamj** [comment]
I believe this issue happens because protocols defined in Swift don't make it through to the Obj-C runtime, where JSONModel would normally inspect them for deserializing into the correct type.

**Comment by billinghamj** [comment]
Ultimately what this comes down to is that we need to provide an alternative method of defining the JSONModel subclasses which should be used in collection types (array/set/dictionary/etc.). Currently JSONModel relies on the use of protocols to achieve this.

Providing an alternative method would be useful for both Obj-C and Swift users, as it allows the header files to be more tidy.

@icanzilb Do you have any opinions on this? I'm not sure we'll use exactly the code in PR #399, but a solution along those lines seems best to me.

**Comment by PerrchicK** [comment]
Sounds good James.
My PR simply adds a compatibility for Swift typed arrays (without harming the existing ObjC usage), just because in Swift you have to define the type of the array. You may change / expand it to be also compatible with other collection.

I've tested it in my project and it solved my problems, that's why I've added this PR.

**Comment by billinghamj** [comment]
@PerrchicK Presumably Swift also requires this for dictionaries?

**Comment by PerrchicK** [comment]
Yes @billinghamj, dictionaries in Swift also required to be strongly-typed, if I understood your question.

https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/CollectionTypes.html#//apple_ref/doc/uid/TP40014097-CH8-ID113

**Comment by AngleZhou** [comment]
+1

**Comment by billinghamj** [comment]
While it's important for us to know that people are interested in this issue, +1 comments aren't helpful.

As an alternative, **please hit the "Subscribe" button on the right of this page** - it shows us a count of how many people 'voted' for the issue.

**Comment by AngleZhou** [comment]
I am using Swift and I cannot transform an array of dictionaries into model. The definition is as follows:
class DJTweet: JSONModel {
    var content: String?
    var uuid: String?
    var updatedAt: String?
    var createdAt: String?
}
class DJTweetsModel: JSONModel {
    var data:[DJTweet]?
    
    override class func protocolForArrayProperty(propertyName: String!) -> String! {
        return "DJTweet"
    }
    override class func propertyIsOptional(propertyName: String!) -> Bool {
        return true
    }
    override class func keyMapper() -> JSONKeyMapper {
        return JSONKeyMapper.mapperFromUnderscoreCaseToCamelCase();
    }
}
I cannot define a protocol the same name as class DJTweet, because it's duplicate. So I tried as above, and it crashes and log:
exception 'Bad property protocol declaration', reason: '<DJTweet> is not allowed JSONModel property protocol, and not a JSONModel class.'

I think this problem can be resolved by allowing collection model subclasses to be defined without using protocols.

**Comment by icanzilb** [comment]
@billinghamj sorry I noticed just now you were asking my input on this. Swift works in a different way than ObjC - JSONModel partially working with Swift classes is pure luck and not something we can rely on. If you start adding features to support Swift classes you will be forced by people reporting problems to keep doing that and as I see ObjC/Swift, Swift is never supposed to have the same reflection superpowers as ObjC does.

**Comment by billinghamj** [comment]
@icanzilb Indeedy, but basically I think we should consider adding the ability to define the JSONModel behavior without reflection. The key thing is doing so in a consistent and scalable way (since as you say, Swift may gradually provide less data to the Obj-C runtime over time).

My inclination is to allow something like defining a static method which can configure the model, e.g.:

```objc
+ (void)configureModel:(JSONModelConfig *)config
{
  [super configureModel:config];

  config[@"collection"].itemType = [MyType class];
  // etc.
}
```

I could also see an approach like this being useful in Obj-C to allow for cleaner header files, without the JSONModel protocol directives.

However, I think that this process would need to be done carefully and gradually, to ensure we don't create something which ends up being more cumbersome in the future.

**Comment by billinghamj** [comment]
@AngleZhou - Swift will not work with collection types until this issue is resolved. For now, if you want to use JSONModel, you'll have to define your models in Obj-C.

**Comment by AngleZhou** [comment]
Thanks for the suggestion. I have switched to jastor, which is less powerful than JSONModel, but works in Swift. Jastor requires a static method, similar to your inclination, to provide the class type as follows:
```
+ (Class)categories_class {
    return [ProductCategory class];
}
```
Hope JSONModel can support this feature soon.

**Comment by billinghamj** [comment]
@AngleZhou The main problem with that kind of approach is it would cause a large number of class methods which would become messy.

Hence combining all the different methods of configuring JSONModel (including perhaps all the existing ones?) would hopefully be a lot tidier.

**Comment by over140** [comment]
`		@property (strong, nonatomic) NSArray<NSDictionary *> *apps;
		@property (strong, nonatomic) NSArray<NSDictionary *> *users;
`
In swift:
`
                   self.users = users.map{ User(dictionary: $0 ) }
                   self.apps = apps.map{ App(dictionary: $0 ) }
`

**Comment by billinghamj** [comment]
This is now resolved in JSONModel v1.3.0. Implement `+ classForCollectionProperty:` as follows:

```objc
+ (Class)classForCollectionProperty:(NSString *)propertyName
{
  if ([propertyName isEqualToString:@"foo"]
    return [MyFooModel class];

  return nil;
}
```

As such, JSONModel should now work reasonably well with Swift, but I will continue to make improvements around this.

I apologize for the delay in getting this resolved.
